### PR TITLE
feat(parsers): port kimi_k2 to the v2 parser

### PR DESCRIPTION
**Pull request by keivenchang**
#### Overview:

This PR ports Kimi K2 only to the v2 parser, preserving reasoning, text, and tool-call order. It also refreshes live-version Dynamo captures and makes fixture extraction safe across concurrent worktrees.

#### Details:

- Shares one Kimi K2 grammar between the tool-only path and v2 parser.
- Covers native and guided tool calls across every valid UTF-8 split, malformed boundaries, and EOF recovery.
- Keys fixture generations by shard identity and selects the live-version capture.

#### Verification:

GitHub Actions pass at `e71ace59`.

#### Where should the reviewer start?

`parsers/v2/src/unified/kimi_k2.rs`, then `conformance/utils/src/extract_fixtures.py`

Closes [DIS-2700](https://linear.app/nvidia/issue/DIS-2700)

/coderabbit profile chill

### CODE REVIEW

**Comment by rmccorm4** [comment]
Looks like some tests failing: https://github.com/ai-dynamo/frontend-crates/actions/runs/32172497284/job/95826711783?pr=191#step:10:536

File: parsers/v2/src/unified/mod.rs [hunk 1]
```diff
@@ -2084,6 +2085,7 @@ macro_rules! unified_registry {
 
 unified_registry! {
     "qwen3" | "qwen3_coder" => qwen3::qwen3_unified,
+    "kimi_k2"               => kimi_k2::kimi_k2_unified,
```
[the remark this answers is not in the document]
Comment (keivenchang) [reply parsers/v2/src/unified/mod.rs:2500]: Already fixed on this head -- `conformance/utils/src/parser_families.yaml:185` sets `native: true` for `kimi_k2` (that's the PR's own one-line change, `native: false -> true`). No further action needed.

File: parsers/v2/src/unified/mod.rs [hunk 2]
```diff
@@ -124,13 +129,50 @@ pub enum UnifiedToolOutputMode {
...
+    /// - A call already on the wire cannot later become recovery text. If a second
+    ///   argument alias appears after the call went out, it is logged, not undone.
+    /// - Recovery is PER CALL. One invalid element of an array surfaces as text on
+    ///   its own; the valid elements around it still dispatch. Under
+    ///   [`RecoverAsText`](Self::RecoverAsText) the whole array would have voided.
+    ///
+    /// What it keeps: the commit point requires the argument value to open with
+    /// `{`, so `null`, a string, a number, an array, and a parameterless call never
+    /// reach the wire early and are still judged by the buffered path.
+    StreamBestEffort,
```
[the remark this answers is not in the document]
Comment (keivenchang) [reply parsers/v2/src/unified/mod.rs:176]: This hunk isn't part of this PR's diff. `git diff <merge-base> HEAD -- parsers/v2/src/unified/mod.rs` only adds `pub mod kimi_k2;` and one `unified_registry!` row; `InvalidGuidedPayloadPolicy::StreamBestEffort` predates this PR and already ships on `main`. GitHub already flags this thread `isOutdated`. Resolving as out of scope for this PR.

File: parsers/v2/src/unified/mod.rs [hunk 3]
```diff
@@ -1705,61 +2005,104 @@ impl GuidedState {
         Ok(calls
             .into_iter()
             .enumerate()
-            .map(|(tool_index, (name, arguments))| {
+            .map(|(tool_index, call)| {
```
[the remark this answers is not in the document]
Comment (keivenchang) [reply parsers/v2/src/unified/mod.rs:2153]: Same as the sibling thread above -- `GuidedState::finish` isn't touched by this PR's diff (confirmed via `git diff <merge-base> HEAD`). Pre-existing code on `main`, already flagged `isOutdated`. Resolving as out of scope for this PR.

File: parsers/v2/src/unified/guided_cursor.rs [hunk 4]
```diff
@@ -0,0 +1,724 @@
...
+                    self.element.in_args = false;
+                    self.maybe_commit(out);
+                    self.flush(payload, cut, out);
+                } else if closes_element {
+                    // A name that closed AFTER its arguments only becomes committable
+                    // here, so commit before the final flush or its bytes are lost
+                    // with the element.
+                    self.maybe_commit(out);
+                    self.flush(payload, at, out);
+                    self.finish_element();
```
[the remark this answers is not in the document]
Comment (keivenchang) [reply parsers/v2/src/unified/guided_cursor.rs:351]: `guided_cursor.rs` isn't touched anywhere in this PR's diff (`git diff <merge-base> HEAD --stat` has no entry for it). Pre-existing code on `main`, already flagged `isOutdated`. Resolving as out of scope for this PR.

File: parsers/v2/src/unified/kimi_k2.rs [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+/// Build the Kimi K2 unified parser for one stream.
+pub(crate) fn kimi_k2_unified(tools: &[Tool]) -> Box<dyn UnifiedParser> {
+    Box::new(ScannerUnified::new(kimi_k2_scanner(tools).with_reasoning(
+        ReasoningSpec {
+            start: REASONING_START,
+            end: REASONING_END,
+            // Kimi emits its own `<think>`; the template does not pre-fill one,
+            // so the stream starts in visible content (policy P5).
+            forced_start: false,
```
[the remark this answers is not in the document]
Comment (keivenchang) [reply parsers/v2/src/unified/kimi_k2.rs:29]: Investigated this one properly, and it doesn't have a valid fix in this crate. Built and tested a `forced_start: true` fix (plus the shared-scanner change the suggested fix needs to not regress every non-reasoning tool call). It passed its own regression tests AND all existing tests, but `cargo test -p dynamo-conformance-fixtures-v2 --test unified_parity` -- the authored golden oracle, not captured from any implementation -- showed 9 NEW divergences, every one a `starting_state=None` (default) case where the golden oracle expects plain non-reasoning text (`UNIFIED.text_only.kimi_k2`, `UNIFIED.text_before_tool.kimi_k2`, etc.) and the fix produced `Reasoning` instead. That's the oracle's own ground truth saying Kimi's default must NOT be reasoning-prefilled -- the opposite of what this fix assumed. Every explicit `starting_state=Reasoning` golden case (`prefilled_reasoning_with_tool` and friends) already passes unmodified, so the override path this finding worried about already exists and already works. Reverted the fix; golden oracle is back to its pre-existing 6 divergences (all unrelated guided-JSON bare-opener cases, not this one). If there's a real gap here it's in whatever caller decides `starting_state` per request for a genuinely-prefilled K2.5 backend, which lives outside this crate. Resolving.

File: parsers/v2/src/unified/kimi_k2.rs [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+use crate::tool_calling::scan::ReasoningSpec;
+use crate::tool_calling::traits::Tool;
+use crate::unified::{ScannerUnified, UnifiedParser};
+
+const REASONING_START: &str = "<think>";
+const REASONING_END: &str = "</think>";
+
+/// Build the Kimi K2 unified parser for one stream.
+pub(crate) fn kimi_k2_unified(tools: &[Tool]) -> Box<dyn UnifiedParser> {
+    Box::new(ScannerUnified::new(kimi_k2_scanner(tools).with_reasoning(
```
[the remark this answers is not in the document]
Comment (keivenchang) [reply parsers/v2/src/unified/kimi_k2.rs:23]: Fixed in e0d992e58. Added a `tool_call_id`/`reset` hook to `InvokeEmitter` (no-op default, so every other family is unaffected) and had `K2Emitter` record `ToolCallResponse::id` -- the native `functions.NAME:IDX` id the v1core parser already extracts -- per `tool_index`, exposed through `ScannerUnified::tool_call_id`. Covered by two new tests (`tool_call_id_preserves_the_native_functions_name_idx_id`, `tool_call_id_does_not_leak_across_reset`), both verified red-before-green. Golden oracle unaffected (still exactly the 6 pre-existing divergences, none involving this).

File: parsers/v2/src/tool_calling/kimi_k2.rs [hunk 7]
```diff
@@ -35,14 +35,206 @@
...
+            return None;
+        }
+        return Some(end);
+    };
+    // From here the shape IS a JSON-argument invoke, so ownership of the
+    // closer search transfers to the JSON boundary (`I7`) -- never fall back
+    // to a raw literal search of the (possibly still-streaming) buffer,
+    // which would re-match a `call_end`-looking byte sequence still sitting
+    // inside the not-yet-closed argument string.
+    let json_end = args_at + json_value_end(&text[args_at..])?;
```
[the remark this answers is not in the document]
Comment (keivenchang) [reply parsers/v2/src/tool_calling/kimi_k2.rs]: Confirmed and fixed in f24162f81. Verified the divergence directly before fixing: for `<|tool_call_argument_begin|>not-json-at-all<|tool_call_end|>`, `try_tool_call_parse_kimi_k2` (batch mode) recovers the call with `arguments: "not-json-at-all"` (its own raw-string fallback), but the streaming boundary finder's `json_value_end(...)?` was propagating `None` and dropping the whole invoke before it ever reached that fallback -- a real stream/batch parity break. Fixed per the suggested direction: when `json_value_end` returns `None`, only at true EOF (`flush`) search for a literal `call_end` (bounded against an intervening `call_start`, matching the pattern already used elsewhere in this function) and let the raw text through to the typing layer's own recovery; mid-stream it still waits, since `None` there usually just means "not balanced yet" (e.g. a still-open string), not malformed -- confirmed this doesn't reopen the I7 embedded-marker case with a dedicated regression test plus an independent audit. New test: `malformed_non_json_arguments_still_ship_the_call_instead_of_vanishing`.

File: parsers/v2/src/tool_calling/kimi_k2.rs [hunk 8]
```diff
@@ -35,14 +35,254 @@
...
+        Some(0) => NativeId::None, // `:` immediately followed by a non-digit
+        Some(d) => NativeId::Complete(ident_len + 1 + d),
+        None if after_colon.is_empty() => {
+            if flush {
+                NativeId::None
+            } else {
+                NativeId::Pending
+            }
+        }
+        None => NativeId::Complete(ident_len + 1 + after_colon.len()),
```
[the remark this answers is not in the document]
Comment (keivenchang) [reply parsers/v2/src/tool_calling/kimi_k2.rs:116]: Investigated and this doesn't reproduce as a wrong final result. Traced why: the `Complete` arm this flags always coincides with the digit run reaching the exact end of the currently-buffered text (since `after_colon` in that arm is a suffix that consumes to the buffer's end by construction), which means the computed boundary always equals `text.len()`, which means `remainder` (`&text[end..]`) is always the empty string at that point. The very next check in the function (`remainder.len() < marker.len() && marker.starts_with(remainder)`) always evaluates true for an empty remainder against both `ARGUMENT_BEGIN` and `CALL_END` (every string is a prefix-match target of an empty needle), so it unconditionally defers to more input arriving whenever `!flush` -- the exact split you're describing (`functions.NAME:1` then `0` in the next chunk) never actually commits early. Verified directly: pushed `...functions.get_weather:10...` split right after the `1`, before the `0` streamed in -- the full two-digit id resolves correctly, same result as a single push. No code change; leaving this thread open only to you to close if my repro doesn't match what you had in mind.

File: parsers/v2/src/tool_calling/scan.rs [hunk 9]
```diff
@@ -134,6 +134,52 @@ where
...
+                escape = true;
+            } else if c == '"' {
+                in_string = false;
+            }
+            continue;
+        }
+        match c {
+            '"' => in_string = true,
+            '{' | '[' => depth += 1,
+            '}' | ']' => {
```
[the remark this answers is not in the document]
Comment (keivenchang) [reply parsers/v2/src/tool_calling/scan.rs:182]: Fixed in 4c389010: json_value_end now tracks the expected closer type, with mismatched-bracket regressions including {].

File: parsers/v2/src/tool_calling/kimi_k2.rs [hunk 10]
```diff
@@ -35,14 +35,82 @@
...
+    // which would re-match a `call_end`-looking byte sequence still sitting
+    // inside the not-yet-closed argument string.
+    let json_end = args_at + json_value_end(&text[args_at..])?;
+    if let Some(rel) = text[json_end..].find(CALL_END) {
+        return Some(json_end + rel + CALL_END.len());
+    }
+    // Best-effort recovery (`UNIFIED.5.b`, policy P2 sibling): the argument
+    // body is syntactically complete but the model stopped before emitting
+    // the closer. Only at true EOF -- otherwise wait for more input.
+    flush.then_some(json_end)
```
[the remark this answers is not in the document]
Comment (keivenchang) [reply parsers/v2/src/tool_calling/kimi_k2.rs]: Fixed in 4c389010: missing-call-end EOF recovery now rejects either intervening Kimi section-end variant, with plural and singular regressions.

File: conformance/utils/src/extract_fixtures.py [hunk 11]
```diff
@@ -232,17 +267,65 @@ def main():
...
+        # path). Any other errno -- permission, I/O, disk full, cross-device
+        # EXDEV -- is a real failure with nothing to reconcile against and
+        # must propagate with its original cause, not be swallowed here.
+        if exc.errno not in RENAME_DEST_EXISTS_ERRNOS:
+            raise
+        # Even a "destination exists" error must be verified, not trusted:
+        # the directory occupying this name could be a stray/partial one
+        # from manual intervention or a bug, not a genuine prior publish.
+        published = read_state(snap_dir)
+        if not (snap_dir.is_dir() and published.get("shards") == pinned_shards):
```
[the remark this answers is not in the document]
Comment (keivenchang) [reply conformance/utils/src/extract_fixtures.py]: Fixed in 4c389010: full refresh publishes a new immutable generation and readers resolve the newest generation; the focused refresh tests cover both behaviors.

File: parsers/v2/src/tool_calling/v1core/xml/kimi_k2_parser.rs [hunk 12]
```diff
@@ -91,22 +97,31 @@ fn find_section_start(
...
-    let mut best: Option<(usize, usize)> = None;
-    for variant in &config.section_end_variants {
-        if let Some(pos) = text[from..].find(variant.as_str())
-            && best.is_none_or(|(bp, _)| pos < bp)
-        {
-            best = Some((pos, variant.len()));
-        }
-    }
-    best
+    find_first_outside_strings(&text[from..], &config.section_end_variants)
```
[the remark this answers is not in the document]
Comment (keivenchang) [reply parsers/v2/src/tool_calling/v1core/xml/kimi_k2_parser.rs]: Fixed in 4c389010: section-end scanning is bounded call by call, so an unmatched quote in malformed raw arguments cannot hide the real section end or swallow trailing text.

File: parsers/v2/src/tool_calling/kimi_k2.rs [hunk 13]
```diff
@@ -35,14 +35,349 @@
...
+            .iter()
+            .any(|marker| trimmed_after_json.starts_with(marker))
+        {
+            return None;
+        }
+    }
+    // `tool_index == 0` only (see the doc comment above): a later call with
+    // no evidence it ever closes is a malformed shape, not truncation, once
+    // an earlier call in the same response DID close correctly.
+    (flush && tool_index == 0).then_some(json_end)
```
[the remark this answers is not in the document]
Comment (keivenchang) [reply parsers/v2/src/tool_calling/kimi_k2.rs:444]: Fixed in 4c389010: missing-closer EOF recovery now requires the bounded argument body to parse as valid JSON, with a bracket-balanced invalid-body regression.

File: parsers/v2/src/tool_calling/scan.rs [hunk 14]
```diff
@@ -849,8 +985,25 @@ impl<E: InvokeEmitter> WrappedBlockScanner<E> {
...
+                // already known to belong to this invoke) -- not for the
+                // full buffer being scanned. A raw search there mistakes an
+                // embedded copy for the real boundary and drops a
+                // well-formed invoke -- also corrupting every later
+                // `tool_index`, since `next_index` never advances for a
+                // dropped call.
                 if self.spec.drop_invoke_crossing_block_end
-                    && let Some((be_pos, be_len)) = find_first(&self.buffer, &self.spec.block_ends)
+                    && let Some((be_pos, be_len)) =
+                        find_first_outside_strings(&self.buffer, &self.spec.block_ends)
```
[the remark this answers is not in the document]
Comment (keivenchang) [reply parsers/v2/src/tool_calling/scan.rs]: Fixed in 4c389010: malformed arguments now take the literal boundary path with the same intervening-section-end guard in batch and streaming, covered by a parity regression.

File: parsers/v2/src/tool_calling/kimi_k2.rs [hunk 15]
```diff
@@ -35,14 +35,386 @@
...
+        return Some(end);
+    };
+    // From here the shape has a real `argument_begin`, so ownership of the
+    // closer search transfers to the JSON boundary (`I7`) when the argument
+    // body IS balanced JSON -- never fall back to a raw literal search of
+    // the (possibly still-streaming) buffer in that case, which would
+    // re-match a `call_end`-looking byte sequence still sitting inside the
+    // not-yet-closed argument string.
+    let after_args = &text[args_at..];
+    let Some(json_len) = json_value_end(after_args) else {
```
[the remark this answers is not in the document]
Comment (keivenchang) [reply parsers/v2/src/tool_calling/kimi_k2.rs]: Fixed in 4c389010: streaming validates JSON immediately after json_value_end before any balanced-body branch can consume the candidate.

File: parsers/v2/src/tool_calling/v1core/xml/kimi_k2_parser.rs [hunk 16]
```diff
@@ -91,22 +97,115 @@ fn find_section_start(
...
+        let section_end_is_first = match (next_section_end, next_arg_begin_start) {
+            (Some((end_pos, _)), Some(arg_pos)) => end_pos < arg_pos,
+            (Some(_), None) => true,
+            (None, _) => false,
+        };
+        if section_end_is_first {
+            return next_section_end;
         }
+        let arg_pos = next_arg_begin_start? + config.argument_begin.len();
+        cursor = match json_value_end(&region[arg_pos..]) {
```
[the remark this answers is not in the document]
Comment (keivenchang) [reply parsers/v2/src/tool_calling/v1core/xml/kimi_k2_parser.rs]: Fixed in 4c389010: batch find_section_end applies the same early JSON-validity gate and routes invalid bodies through the guarded malformed-argument path.

File: parsers/v2/src/tool_calling/v1core/xml/kimi_k2_parser.rs [hunk 17]
```diff
@@ -307,11 +420,38 @@ fn parse_section_block(
...
+        // capture used, whenever the body is clean JSON immediately followed
+        // by the real closer; fall back to the lazy capture for genuinely
+        // malformed/non-JSON argument bodies, which still need the
+        // permissive raw-string path (`serde_json::from_str` below falls
+        // back to the raw text on a parse error either way).
+        let arguments_raw = cap
+            .name("arguments")
+            .and_then(|m| {
+                let args_start = m.start();
+                let json_end = args_start + json_value_end(&block[args_start..])?;
```
[the remark this answers is not in the document]
Comment (keivenchang) [reply parsers/v2/src/tool_calling/v1core/xml/kimi_k2_parser.rs]: Fixed in 4c389010: the I7 override now replaces the lazy boundary only when the candidate parses as valid JSON, with invalid and valid sibling regressions.

File: conformance/utils/src/extract_fixtures.py [hunk 18]
```diff
@@ -232,17 +337,127 @@ def main():
...
+        shutil.rmtree(str(tmp_dir))
+    print(f"Extracting {len(shards)} shard(s) into {tmp_dir}", file=sys.stderr)
     for s in shards:
-        extract_tarball(shard_file(s), snap_dir, verbose=args.verbose)
-    write_state(snap_dir, pin, shards)
+        extract_tarball(shard_file(s), tmp_dir, verbose=args.verbose)
+    write_state(tmp_dir, pin, shards)
+
+    try:
+        tmp_dir.rename(snap_dir)
```
[the remark this answers is not in the document]
Comment (keivenchang) [reply conformance/utils/src/extract_fixtures.py]: Fixed in 56981f56. `--full-refresh` now chooses the next generation from the highest valid published generation even when generation 0 is missing. The same locked publication transition also serializes final link retargeting, and current Python consumers resolve the immutable snapshot path rather than the compatibility links. Regression coverage includes missing-base generations 1 and 3 plus a forced concurrent-refresher interleaving.

File: parsers/v2/src/tool_calling/kimi_k2.rs [hunk 19]
```diff
@@ -35,14 +35,419 @@
...
+    // has no fallback for a missing `call_end` regardless of what follows
+    // it, so recovering here would ship a call batch mode drops -- exactly
+    // the divergence `conformance_toolcalling_batch_via_stream` caught.
+    if flush {
+        let trimmed_after_json = after_json.trim_start();
+        if [SECTION_END_PLURAL, SECTION_END_SINGULAR]
+            .iter()
+            .any(|marker| trimmed_after_json.starts_with(marker))
+        {
+            return None;
```
Comment (indrajit96) [inline parsers/v2/src/tool_calling/kimi_k2.rs:434]: Returning None here correctly prevents synthesizing a missing call_end, but it also makes `WrappedBlockScanner::drain` take its `flush && invoke_end == None` branch, which clears the entire remaining buffer.

For example:
 ```
<|tool_calls_section_begin|><|tool_call_begin|>functions.lookup:0<|tool_call_argument_begin|>{}<|tool_calls_section_end|>Visible answer
```
 Batch parsing drops the malformed call and preserves Visible answer, while streaming drops both because the buffer containing section_end and the suffix is cleared at EOF. 
The correct result is calls: [] and `normal_text: "Visible answer".`

Please preserve/signal the detected section boundary so the scanner can drop through section_end, leave block mode, and continue draining the suffix as visible text, rather than treating the whole remainder as an unbounded incomplete invoke.
A batch-via-stream regression with trailing text would pin this behavior.

### Suggested regression test

Add it beside the existing `mismatched_fences_drop_the_call_instead_of_recovering_it` (https://github.com/ai-dynamo/frontend-crates/blob/e71ace598bf8909184b2a438267ce04614881497/parsers/v2/src/tool_calling/kimi_k2.rs#L872-L893) test. 
That test currently ends immediately after section_end, so it verifies only that the call is dropped and cannot detect suffix loss.

Conceptually:
```
  #[test]
  fn mismatched_fences_preserve_text_after_section_end() {
      let out = parse_chunks(
          &weather_tools(),
          &[
              "<|tool_calls_section_begin|>\
               <|tool_call_begin|>functions.get_weather:0\
               <|tool_call_argument_begin|>{}\
               <|tool_calls_section_end|>\
               Visible answer",
          ],
      );

      assert!(out.calls.is_empty());
      assert_eq!(out.normal_text, "Visible answer");
  }
```
The important fix requirement is not to recover the malformed tool call—it must remain dropped. Only the content after the real outer section boundary should survive.

File: parsers/v2/src/tool_calling/kimi_k2.rs [hunk 20]
```diff
@@ -35,14 +35,419 @@
...
+        // trust it only once nothing more is coming (`flush`). Same class of
+        // bug either commit destroyed -- reading it early made the outcome
+        // depend on the chunk boundary instead of the bytes.
+        if !flush && remainder.starts_with(CALL_END) {
+            return None;
+        }
+        if !flush
+            && [ARGUMENT_BEGIN, CALL_END]
+                .iter()
+                .any(|marker| remainder.len() < marker.len() && marker.starts_with(remainder))
```
Comment (indrajit96) [inline parsers/v2/src/tool_calling/kimi_k2.rs]: This partial-marker check uses remainder verbatim, but the Kimi batch grammar permits \s* between NAME:IDX and argument_begin. If a chunk ends after the whitespace, for example:
```
chunk 1: <|tool_calls_section_begin|><|tool_call_begin|>functions.search:0
chunk 2: <|tool_call_argument_begin|>{"query":"rust"}<|tool_call_end|><|tool_calls_section_end|>
```
then `remainder == " "`, so neither marker starts with it and this function returns Some(end). 
The scanner consumes the call_begin and native ID before argument_begin arrives; K2Emitter cannot parse that header-only span, and chunk 2 no longer contains an invoke opener, so the otherwise-valid call is silently lost. 
Passing the same concatenated bytes in one chunk succeeds.
 
Can we make the holdback decision after ignoring permitted leading whitespace while retaining the original buffered bytes and add an every-split regression using whitespace before argument_begin. 
The complete and split delivery schedules should both emit the search call.

### Suggested Fix :

For the specific case, both marker checks need to reason over the whitespace-trimmed suffix:

```
  let remainder = &text[end..];
  let structural_remainder = remainder.trim_start();

  if !flush && structural_remainder.starts_with(CALL_END) {
      return None;
  }

  if !flush
      && [ARGUMENT_BEGIN, CALL_END].iter().any(|marker| {
          structural_remainder.len() < marker.len()
              && marker.starts_with(structural_remainder)
      })
  {
      return None;
  }
```
  
The original remainder should remain buffered. trim_start() is only for deciding whether a delimiter may still be arriving; the parser should not independently emit or discard that whitespace.

This handles:
```
  " "                         // whitespace only
  "   "                       // several whitespace bytes
  " <|tool_call_arg"          // whitespace plus partial argument marker
  "\n<|tool_call_argument_"   // newline plus partial marker
```
### Suggested regression test :

The best location is beside a_real_native_id_is_still_recognized_across_every_split (https://github.com/ai-dynamo/frontend-crates/blob/e71ace598bf8909184b2a438267ce04614881497/parsers/v2/src/unified/kimi_k2.rs#L539-L556).

That existing test sweeps every split point but has no whitespace between the ID and argument marker, so an empty remainder protects it.

A regression could be:
```
  #[test]
  fn whitespace_before_argument_marker_is_split_invariant() {
      let input = concat!(
          "<|tool_calls_section_begin|>",
          "<|tool_call_begin|>functions.get_weather:0 ",
          "<|tool_call_argument_begin|>{\"city\":\"Paris\"}",
          "<|tool_call_end|>",
          "<|tool_calls_section_end|>"
      );

      let want = vec![UnifiedEvent::ToolCall {
          name: "get_weather".into(),
          arguments: serde_json::json!({"city": "Paris"}),
      }];

      for (i, got) in assemble_at_every_split(&tools(), input)
          .into_iter()
          .enumerate()
      {
          assert_eq!(
              got, want,
              "valid whitespace-delimited call changed at split byte {i}: {got:?}"
          );
      }
  }
```

**Review by indrajit96 — APPROVED** [review]
LGTM!
2 corner cases added in comments.

2 seperate followp questions

1. I was curious how are we handling rust tests and conformance test cases here? Are the tests we had for V1 directly flow to V2 and Unified? Ideally we would like to run ALL we have for V1 against V2 and Unified.
2. Let me know when the PR is ready for E2E tests I can run either for all models or ONLY Kimi K2.X depending on your prefrence.

File: parsers/v2/src/tool_calling/kimi_k2.rs (hunk 20 above)
Comment (keivenchang) [reply parsers/v2/src/tool_calling/kimi_k2.rs]: Fixed in dcf57958. Holdback check now decides on `remainder.trim_start()`, not raw `remainder` — `end` is unchanged so the whitespace stays buffered. `whitespace_before_argument_marker_is_split_invariant` covers it.

File: parsers/v2/src/tool_calling/kimi_k2.rs (hunk 19 above)
Comment (keivenchang) [reply parsers/v2/src/tool_calling/kimi_k2.rs:434]: Fixed in dcf57958. `drain` now checks for a real block-end marker before clearing the buffer on `None`, instead of always discarding it. `mismatched_fences_preserve_text_after_section_end` covers it. One pre-existing gap left open: a valid second call before the section-end still gets dropped — separate follow-up.

**Comment by keivenchang** [comment]
Thanks for the review.

The V1 unit tests don't run directly against V2/Unified. We reuse the fixture corpus: `conformance_toolcalling_batch_via_stream` sends every batch fixture's full `model_text` through the V2 streaming parser and compares the calls with V1's captured `expected.dynamo_v1`; `conformance_toolcalling_stream` repeats that check chunk by chunk. Unified has its own separately authored `UNIFIED.*` golden set for call order, chunk invariance, stream/batch parity, and per-stream isolation. It uses the same case shapes as V1 without being generated from the V1 Rust test functions.

So the precise guarantee is that every V1 batch fixture is replayed through V2 streaming and checked against V1's expected output; the V1 test functions themselves are not reused 1:1. Literal 1:1 reuse would be follow-up work.

For E2E, all PR checks and the full workspace suite pass with zero failures. Kimi K2 is ready to run E2E now; I can start with Kimi K2.x or run the full model set, whichever you prefer.
